src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Data;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Autofac;
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
```

Comment:
That's kind of interesting because the transaction manager should be thread-local since there is a separate lifetime scope for each lock operation (i.e. each thread) and nothing should prevent multiple transactions to be handled in parallel (since this is also what happens during requests...).

It seems to me that there is some more involved issue here that a synchronization would just sweep under the rug. Maybe the recent changes regarding session management caused this somehow?